False positive with loop + switch

Hello!

```go
package main

import "fmt"

func main() {
	flag := false
	for i := 0; i < 10; i++ {
		switch {
		case false:
			continue
		case true:
			flag = true
		}
	}

	if flag {
		fmt.Println("hello")
	}
}
```

```
golangci-lint run main.go
main.go:6:2: variable 'flag' is only used in the if-statement (main.go:16:2); consider using short syntax (ifshort)
        flag := false
        ^
```

--- a/pkg/analyzer/analyzer.go
+++ b/pkg/analyzer/analyzer.go
@@ -85,8 +85,9 @@ func (nom namedOccurrenceMap) checkStatement(stmt ast.Stmt, ifPos token.Pos) {
 			nom.checkExpression(a, ifPos)
 		}
 	case *ast.ExprStmt:
-		if callExpr, ok := v.X.(*ast.CallExpr); ok {
-			nom.checkExpression(callExpr, ifPos)
+		switch v.X.(type) {
+		case *ast.CallExpr, *ast.UnaryExpr:
+			nom.checkExpression(v.X, ifPos)
 		}
 	case *ast.ForStmt:
 		for _, el := range v.Body.List {
@@ -157,12 +158,11 @@ func (nom namedOccurrenceMap) checkStatement(stmt ast.Stmt, ifPos token.Pos) {
 			}
 
 			for _, c := range clauses.Body {
-				if est, ok := c.(*ast.ExprStmt); ok {
-					nom.checkExpression(est.X, ifPos)
-				}
-
 				switch v := c.(type) {
 				case *ast.AssignStmt:
+					for _, el := range v.Lhs {
+						nom.checkExpression(el, ifPos)
+					}
 					for _, el := range v.Rhs {
 						nom.checkExpression(el, ifPos)
 					}
@@ -171,6 +171,28 @@ func (nom namedOccurrenceMap) checkStatement(stmt ast.Stmt, ifPos token.Pos) {
 				}
 			}
 		}
+	case *ast.SelectStmt:
+		for _, el := range v.Body.List {
+			clause := el.(*ast.CommClause)
+
+			nom.checkStatement(clause.Comm, ifPos)
+
+			for _, c := range clause.Body {
+				switch v := c.(type) {
+				case *ast.AssignStmt:
+					for _, el := range v.Lhs {
+						nom.checkExpression(el, ifPos)
+					}
+					for _, el := range v.Rhs {
+						nom.checkExpression(el, ifPos)
+					}
+				case *ast.ExprStmt:
+					nom.checkExpression(v.X, ifPos)
+				}
+			}
+		}
+	case *ast.LabeledStmt:
+		nom.checkStatement(v.Stmt, ifPos)
 	}
 }
 
@@ -190,7 +212,7 @@ func (nom namedOccurrenceMap) checkExpression(candidate ast.Expr, ifPos token.Po
 	case *ast.CompositeLit:
 		for _, el := range v.Elts {
 			switch v := el.(type) {
-			case *ast.Ident:
+			case *ast.Ident, *ast.CompositeLit:
 				nom.checkExpression(v, ifPos)
 			case *ast.KeyValueExpr:
 				nom.checkExpression(v.Key, ifPos)
@@ -217,6 +239,8 @@ func (nom namedOccurrenceMap) checkExpression(candidate ast.Expr, ifPos token.Po
 				}
 			}
 		}
+	case *ast.StarExpr:
+		nom.checkExpression(v.X, ifPos)
 	case *ast.IndexExpr:
 		nom.checkExpression(v.X, ifPos)
 		switch index := v.Index.(type) {

--- a/testdata/dummies.go
+++ b/testdata/dummies.go
@@ -12,12 +12,16 @@ func getBool(...interface{}) bool { return false }
 
 func getInt(...interface{}) int { return 0 }
 
+func getChan(...interface{}) chan interface{} { return nil }
+
 type dummyType struct {
 	interf interface{}
 	slice  []interface{}
 }
 
 func getDummy(...interface{}) dummyType { return dummyType{} }
+
+func getDummyPtr(...interface{}) *dummyType { return &dummyType{} }
 
 func getTwoDummies(...interface{}) (dummyType, dummyType) { return dummyType{}, dummyType{} }
 

--- a/testdata/testdata.go
+++ b/testdata/testdata.go
@@ -518,3 +518,52 @@ func notUsed_BinaryExprInAssign_OK() {
 		noOp1(v1)
 	}
 }
+
+func notUsed_ReferenceInSelect_OK() {
+	v := 0
+	if getBool(v) {
+		v = 1
+	}
+	select {
+	case <-getChan(v):
+	}
+}
+
+func notUsed_AssignInSwitch_OK() {
+	y := 100
+	switch {
+	case true:
+		y = 1
+	}
+	if y < 5 {
+	}
+}
+
+func notUsed_PassPointer_OK() {
+	a := getDummyPtr()
+	if a == nil {
+	}
+	noOp1(*a)
+}
+
+func notUsed_Labels_OK() {
+	foo := true
+BREAKOUT:
+	if getBool() {
+		foo = false
+		goto BREAKOUT
+	}
+	if foo {
+		return
+	}
+}
+
+func notUsed_UnnerStruct_OK() {
+	v := getInt()
+	if v == 0 {
+		v = 1
+	}
+	type Wrapper struct{ V int }
+	t := []Wrapper{{v}}
+	noOp1(t)
+}

